fix: use existing serviceAccount when specified

cyon13 opened the pull request:
**What this PR does / why we need it**:

This PR changes how ServiceAccounts are handled for multi-node main and prefill workloads when
serviceAccountName is explicitly specified in LLMInferenceService.

Previously, when a user provided serviceAccountName, KServe created a new managed ServiceAccount
and copied selected fields (labels, annotations, secrets, imagePullSecrets, etc.) from the existing
ServiceAccount into the newly created one. While functional, this behavior was unintuitive, as
users typically expect the explicitly referenced ServiceAccount itself to be used directly.

With this change, when serviceAccountName is specified, KServe will use the existing
ServiceAccount as-is, instead of creating and managing a derived ServiceAccount. This aligns more
closely with user expectations and simplifies ServiceAccount ownership and lifecycle management.

This change is also required for environments such as Azure AKS when using Azure Workload
Identity. In such setups, access to Azure resources (e.g. Azure Blob Storage) depends on a
pre-existing ServiceAccount that is bound to a specific Azure identity. Creating a new
ServiceAccount—even with copied metadata—breaks this identity binding and prevents workloads from
accessing Azure Blob Storage correctly. By directly using the existing ServiceAccount, KServe
properly supports these identity-based access patterns.

When no serviceAccountName is specified, KServe continues to create and manage its own
ServiceAccounts as before.

---

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4903 

---

**Type of changes**
Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

---

**Feature/Issue validation/testing**:

- Verified behavior when `serviceAccountName` is specified for:
  - Multi-node main workload
  - Multi-node prefill workload
  - Scheduler (used by endpointPicker)
  - Single-node PD disaggregation serving
- Confirmed that:
  - No new ServiceAccount is created
  - The existing ServiceAccount is referenced directly across all components
- Verified that when `serviceAccountName` is not specified:
  - KServe-managed ServiceAccounts are still created and reconciled as before
    for main, prefill, scheduler, and single-node workloads
- Added test coverage for single-node ServiceAccount handling to validate
  reuse of existing ServiceAccounts when specified

Manual testing was performed by deploying `LLMInferenceService` manifests with and without
`serviceAccountName` fields set.

- [x] Manual test with existing ServiceAccount
- [x] Manual test without ServiceAccount specified

- Logs  
  (No error or reconciliation failures observed)

---

**Special notes for your reviewer**:

1. This PR intentionally removes the previous behavior of copying fields from an existing
   ServiceAccount into a newly created managed ServiceAccount.
2. Ownership and lifecycle of user-provided ServiceAccounts are now fully left to the user.
3. RBAC reconciliation logic remains unchanged and continues to work with the referenced
   ServiceAccount.

---

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

---

**Release note**:

```release-note
Updated ServiceAccount handling to directly reuse user-specified serviceAccountName instead of creating managed ServiceAccounts.

cyon13 committed:
Use existing ServiceAccount if present instead of creating a new one

Signed-off-by: cyon13 <bonyul@naver.com>

sivanantha321 commented on pkg/controller/v1alpha2/llmisvc/workload_multi_node.go:
```diff
@@ -329,18 +329,27 @@ func (r *LLMISVCReconciler) expectedPrefillMultiNodeLWS(ctx context.Context, llm
 }
 
 func (r *LLMISVCReconciler) reconcileMultiNodeMainServiceAccount(ctx context.Context, llmSvc *v1alpha1.LLMInferenceService, config *Config) error {
+	useExistingServiceAccount := false
```
Can you add a test for this case ?

cyon13 replied:
@sivanantha321 Thanks for the suggestion.
I’ve added unit tests to cover this scenario by explicitly validating the ServiceAccount reconciliation behavior in the following cases:

For the main workload
1. When an existing serviceAccountName is specified, the reconciler uses it as-is and does not create a managed ServiceAccount.
2. When no serviceAccountName is specified and a worker is present, a managed ServiceAccount is created.
3. When no serviceAccountName is specified and the worker is nil, an existing managed ServiceAccount is deleted.

Equivalent test coverage has also been added for the prefill workload, covering the same three cases.

cyon13 committed:
Add unit tests for multi-node main and prefill ServiceAccount reconciliation

Signed-off-by: cyon13 <bonyul@naver.com>

sivanantha321 commented:
@cyon13 Can you help sign off the commit ?

cyon13 committed:
Use existing ServiceAccount for PD disaggregation serving and add single-node ServiceAccount tests

Signed-off-by: cyon13 <bonyul@naver.com>

cyon13 commented:
@sivanantha321 I’ve signed off the commit.
In addition, I updated workload_single_node.go so that PD disaggregation serving follows the same behavior: when serviceAccountName is specified, the controller reuses the existing ServiceAccount instead of creating a managed one. I also added single-node ServiceAccount tests to cover this behavior.

cyon13 commented:
@sivanantha321 The test failures seem unrelated to this change (Go/Python tests).
I’ve updated the branch to the latest base branch.
Could you please help re-run the tests?

cyon13 commented:
@sivanantha321 Since github.com/kserve/kserve/pkg/apis/serving/v1alpha1 was changed to github.com/kserve/kserve/pkg/apis/serving/v1alpha2 in master, I updated the code accordingly to align with the new API version.

cyon13 committed:
reuse existing scheduler service account if present

Signed-off-by: cyon13 <bonyul@naver.com>

cyon13 commented:
@sivanantha321 While testing the endpointPicker functionality, I noticed that the scheduler also uses a ServiceAccount.
To keep the behavior consistent, I updated the scheduler to follow the same logic—reusing an existing ServiceAccount when one is explicitly specified, instead of creating a new managed one.

sivanantha321 commented:
@cyon13 looks like the go tests are failing. Can you look into it ?

cyon13 committed:
Fix ServiceAccount logic causing test failures

Signed-off-by: cyon13 <bonyul@naver.com>

cyon13 commented:
@sivanantha321  I fixed the failing tests and identified two root causes related to ServiceAccount handling.

First, when serviceAccountName is specified, the controller is expected to reuse the specified ServiceAccount instead of creating a new one. However, some existing tests were still asserting the presence of a newly generated ServiceAccount name even though serviceAccountName was explicitly set, which caused test failures.

Second, previously the controller only created a new ServiceAccount when serviceAccountName was empty. This led to an issue when serviceAccountName was specified but had the same value as the ServiceAccount name that KServe would normally generate. In this case, the tests failed because the logic did not correctly handle this overlap.

To resolve this, I updated the logic so that when a user-specified serviceAccountName matches the ServiceAccount name that KServe would generate, the controller follows the managed ServiceAccount creation path. I verified the changes by running `make test`, and all tests passed.

Could you please rerun the tests?

sivanantha321 committed:
Merge branch 'master' into multinode-sa

sivanantha321 commented:
Thanks for addressing all the comments!
/lgtm
/approve
